// @bovine/mucked-up-components/src/ActorHeader/index.tsx
import type { ActorHeaderInfo } from "@bovine/muck-out-types";

import { ActorAvatar } from "../ActorAvatar";
import { headerIdentifierStyle, headerNameStyle } from "./styles";
import { linkStyle } from "../styles";

type ActorHeaderProps = {
  actorInfo: ActorHeaderInfo;
  style?: object;
};

function transformIdentifier(identifier: string): string {
  return identifier.replace("acct:", "@");
}

function ActorHeaderName({ actorInfo }: ActorHeaderProps) {
  const name = actorInfo.name || "- unknown -";

  if (actorInfo.htmlUrl) {
    return (
      <div style={headerNameStyle}>
        <a style={linkStyle} href={actorInfo.htmlUrl}>
          {name}
        </a>
      </div>
    );
  }

  return <div style={headerNameStyle}>{name}</div>;
}

function ActorHeaderIdentifier({ actorInfo }: ActorHeaderProps) {
  return (
    <div style={headerIdentifierStyle}>
      <a style={linkStyle} href={actorInfo.id}>
        {transformIdentifier(actorInfo.identifier)}
      </a>
    </div>
  );
}

function ActorHeader({ actorInfo, style }: ActorHeaderProps) {
  return (
    <header style={{ ...style, display: "flex" }}>
      <ActorAvatar actorInfo={actorInfo} />
      <span>
        <ActorHeaderName actorInfo={actorInfo} />
        <ActorHeaderIdentifier actorInfo={actorInfo} />
      </span>
    </header>
  );
}

export { ActorHeader, type ActorHeaderProps };
