### `docs/Genomesetup.md`

# Genomesetup: Genome Index Preparation

`Genomesetup` sets up genome indices and promoter annotation for following ProActiv, Salmon and DEXSeq analysis.


## Usage
```bash
Genomesetup [options]
```

### Required Arguments
- `--organism`: Genome assembly (e.g., `hg38`, `mm10`).
- `--organism_fasta`: Absolute path to genome FASTA file (e.g., `/tmp/Genomes/GRCh38/genome.fa`). 
- `--genes_gtf`: Absolute path to gene gtf annotation file (e.g., `/tmp/Genomes/GRCh38/genes.gtf`). 

### Optional Arguments
- `-o/--output_dir`: Directory to store genome files (e.g., `genome`).
- `--threads`: Number of CPU threads (e.g., `30`). Default: 30.

### Example
```bash
Genomesetup \
  --organism hg38 \
  --organism_fasta /path/to/your/genome.fa \
  --genes_gtf /path/to/your/genes.gtf \
  -o ./genome \
  --threads 30
```

## Output
```
tree organisms/hg38
organisms/hg38
├── benchmarks
│   ├── copy_fasta_hg38.bmk.txt
│   ├── copy_gtf_hg38.bmk.txt
│   ├── dexseq_prepare_hg38.bmk.txt
│   ├── generate_chrom_sizes
│   │   └── hg38.txt
│   ├── gtf_to_files_hg38.bmk.txt
│   ├── hisat2Index_hg38.bmk.txt
│   ├── proactiv_prepare_hg38.bmk.txt
│   ├── salmon_index_hg38.bmk
│   └── starIndex_hg38.bmk.txt
├── logs
│   └── generate_chrom_sizes
├── organisms
│   └── hg38
│       ├── Annotation
│       │   ├── DEXSeq_flattened_exons.gff
│       │   ├── genes.bed
│       │   ├── genes.symbol
│       │   ├── genes_t2g.tsv
│       │   ├── proActiv_promoter_annotation.rds
│       │   ├── proActiv_promoter.bed
│       │   ├── proActiv_protein_coding.gtf
│       │   ├── proActiv_txdb.sqlite
│       │   └── transcripts.fa
│       ├── Annotation.gtf
│       ├── hg38.chrom.sizes
│       ├── hg38.fa
│       ├── hg38.fa.fai
hg38/hg38.fa
│       ├── main_chroms.txt
│       ├── SalmonIndex
│       │   ├── complete_ref_lens.bin
│       │   ├── ctable.bin
│       │   ├── ctg_offsets.bin
│       │   ├── decoys.txt
│       │   ├── duplicate_clusters.tsv
│       │   ├── info.json
│       │   ├── mphf.bin
│       │   ├── pos.bin
│       │   ├── rank.bin
│       │   ├── refAccumLengths.bin
│       │   ├── reflengths.bin
│       │   ├── refseq.bin
│       │   ├── seq.bin
│       │   └── versionInfo.json
│       └── STARIndex
│           ├── chrLength.txt
│           ├── chrNameLength.txt
│           ├── chrName.txt
│           ├── chrStart.txt
│           ├── Genome
│           ├── genomeParameters.txt
│           ├── SA
│           └── SAindex


```

- **`benchmarks`** - Benchmarking data for pipeline steps, including runtime and memory usage per sample (e.g., `benchmarks/copy_fasta_hg38.bmk.txt`).
- **`logs`** - Detailed log files for individual pipeline steps (e.g., `logs/copy_fasta_hg38.log`).

-**`organisms/hg38`**: Directory for input, annotation, and index.

# Setup files
- **`hg38.fa`**: Genome FASTA file of DNA sequence in hg38 human reference genome.
- **`hg38.fa.fai`**: FASTA Index file generated by `samtools`, enables fast access of sequences in `hg38.fa`.
- **`hg38.chrom.sizes`**: Genome chromosome size file of chromosome name and lenths.
- **`Annotation.gtf`**: Gene annotation file in GTF format for alignment and quantification, contains information about chromosome name, gene name, gene coordinates, strandness, geneid, gene type, transcript id, transcript name, transcript type, exon coordinates.
- **`main_chroms.txt`**: A list of chromosome names.

# Annotation
- **`Annotation`**: Annotation files for downstream analysis. Includes:
  - **`DEXSeq_flattened_exons.gff`**: Flattened exon bin annotation in gff format, includes coordinate of exon bin, gene ids the exon bin overlaps with, strandness, and the exonic_part_number of the bin. Used for DEXSeq exon bin counting (e.g., in Altpromoterflow pipeline).
  - **`genes_t2g.tsv`**: Transcript to gene mapping in tsv format, includes transcript id, gene id, and gene name. Used for transcript-level aggregation to gene-level (e.g., tximport).
 - **`genes.bed`**: Readable BED format file with chromosome, chromosomeStart, chromosomeEnd, transcript name, score, strandness, thickStart, thickEnd, itemRgb, blcokCount, blockSizes, blockStarts (where block here is exon). 
 - **`genes.symbol`**: TSV file of gene id and gene symbol.

  ## Promoter Annotation
 - **`proActiv_protein_coding.gtf`**: GTF file containing only protein-coding transcripts, used to generate the promoter annotation.
  - **`proActiv_promoter_annotation.rds`**: Promoter annotation for downstream promoter analysis, clustering of first exon and map onto promoter ID. S4 Promoter annotation file with one class, `PromoterAnnotation` and three slots. 
    - Slot `intronRanges` contains seqnames <Rle>, ranges <IRanges>, strand <Rle>, and two metadata columns INTRONID <integer>, TXNAME <CharacterList>. 
    - Slot `promoterIdMapping` includes transcriptName, promoterId, and geneId. 
    - Slot `promoterCoordinates` contains seqnames  <Rle> , ranges <IRanges>, strand <Rle>, and four metadata columns promoterId <integer> , internalPromoter <logical> , firstExonEnd <integer>, intronId <IntegerList>.
  - **`proActiv_promoter.bed`**: BED file listing promoter regions, ids, score, strandness, used for visualization or overlap analysis.
  - **`proActiv_txdb.sqlite`**: A TxDb object (SQLite) used by proActiv to retrieve genomic structures (exon, transcript, gene). It is S4 type object with Class `TxDb` and slotName `".xData"`. 
    - `transcripts(txdb)` contains seqnames <Rle>, ranges <IRanges>, strand <Rle>, and two metadata columns tx_id <integer>, tx_name <character>. 
    - `exons(txdb)` contains seqnames <Rle>, ranges <IRanges>, strand <Rle>, and one metadata column exon_id <integer>.
    - `genes(txdb)` contains seqnames <Rle>, ranges <IRanges>, strand <Rle>, and one metadata column gene_id <character>.


# Index
- **`STARIndex`**: Building reference index for STAR (Spliced Transcripts Alignment to a Reference) aligner. 
  - Includes chromosome info files  (e.g., `chrLength.txt`, `chrName.txt`, `chrNameLength.txt`, `chrStart.txt`) for mapping from genome onto chromosome position.
  - Includes parameters (e.g.,`genomeParameters.txt`) which includes genome version, FASTA file, Suffix Array index k-mer length, chromosome bin bits, sparsity of Suffix Array, splice junction overhang, GTF exon's feature name, transcript id from gtf, gene id from gtf, save splice junction format.
  - Creats binary files required for splice-aware RNA-seq alignment (e.g., `SA`, `SAindex`, `Genome`). 

- **`SalmonIndex`**: Building decoy-aware transcriptome index for Salmon.
  - **`decoys.txt`**: List of decoy (non-coding or genomic) sequences included to reduce spurious mappings. It is a list of chromosome names，indicating the chromosomes whose sequences are added from `hg38.fa` into decoy.
  - **`duplicate_clusters.tsv`**: Mapping of redundant transcripts grouped together in tsv file. Contains two columns RetainedRef and DuplicateRef.
  - **`info.json`**, **`versionInfo.json`**: Metadata for the Salmon index. Contains (e.g., k-mer length, number of kmers, sesquence length, SeqHash, NameHash, DecoySeqHash, number of decoys, index version).
  - **`seq.bin`**, **`refseq.bin`**: Encoded transcript and reference sequences for reading k-mer from transcriptome during quasi mapping.
  - **`ctable.bin`**, **`rank.bin`**, **`mphf.bin`**, **`ctg_offsets.bin`**, **`pos.bin`**: Binary files used for quasi-mapping and transcript quantification. Map k-mer onto index, store position on transcript, rank and select to speed up mapping, additional info on position and offset.  
  - **`complete_ref_lens.bin`**, **`refAccumLengths.bin`**, **`reflengths.bin`**: Length metadata used during abundance estimation. Transcript length for offset calculation for read length in downstream TPM or NumReads.



## Notes
- Ensure the FASTA file has chromosome names prefixed with `chr` (e.g., `chr1`).
- Use this output folder as `--genome_dir` in `Snakealtpromoter.py`.
